source/codegen/format_mi_metadata.py

```diff
@@ -104,8 +104,8 @@ def _format_mi_metadata(metadata_dir: str):
                 f"# This file is generated from {api_name} API metadata version {api_version}"
             )
             pretty_metadata += "\r\n"
-            pretty_metadata += f"{metadata_name} = {pretty(actual_metadata)}"
-            with codecs.open(f"{path}{metadata_name}.py", "w", "utf-8-sig") as temp:
+            pretty_metadata += f"{metadata_name} = {pretty(actual_metadata)}\r\n"
+            with io.open(f"{path}{metadata_name}.py", "w", encoding="utf-8", newline="\n") as temp:
```

Comment:
Note: This fixes up the encoding to match what hapigen mutation / format uses. Also, the newlines in the hapigen output for the bulk of the exported files are just `\n` so we set that here as well.

Reply:
I'm surprised by this because both git and python have special features to give platform newlines. In git, the repo version will have unix newlines that are converted to platform newlines on checkout (by default). And python's [default file I/O is "universal newlines"](https://docs.python.org/3/library/functions.html#open-newline-parameter) where `\n` is always written out as a platform newline.

So I'd expect:
1. There should never be a `\r\n` in the python code. The line above should end with `\n`.
   * It looks like some of the existing code outside of this diff should be changed as well.
3. The `newline` param to `io.open` should not be used. The default `None` is correct ("universal newlines").

Are you sure that doesn't work? If the problem is that it doesn't matches `hapigen` (a) that's still surprising and (b) the `git diff` call is already ignoring line separators. So, it seems that this code is both wrong and does not match hapigen. Might as well be right and not match hapigen.

Reply:
Yeah, I was getting a bit confused about all the newline stuff. Changed all the `\r\n` in this file to just `\n` and got rid of the `newline` arg to `io.open` so it just relies on the default being `None`. That at least makes it right in this script.

That should work and I updated the validation .yml file to just call `git diff --exit-code`. It doesn't need to ignore end lines like I initially thought.

As for the hapigen output, at least right now it will have `\r\n` for three lines (corresponding to two comment lines at the top and the ending blank line the .mako file has). It then uses `\n` for all the lines corresponding to the pretty formatted dictionary. You can see that .mako [here](https://dev.azure.com/ni/DevCentral/_git/ni-central?path=/src/hapigen/hapigen_grpc_device_plugin/source/templates/grpc_device.py.mako).